File: explore/string/move_zeroes.py
# Given an array of nums, write a function to move all the 0s to the end of it while maintaining the relative order of the other elements. Do it in place with O(1) space.
# Slow and fast runner? Slow runner will keep place while other one will iterate. Swap on nonzero encounters.
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        # slow runner keeps track of swap position
        slow = 0

        # fast runner encounters a nonzero and swaps with the slow runner value
        for i in range(len(nums)):
            if nums[i] != 0:
                nums[i], nums[slow] = nums[slow], nums[i]
                slow += 1

File: golang/move_zeroes.go
// given an integer array, move all 0's to the end while maintaining relative order of nonzero elements
// use two pointer method and swap on nonzero occurrence
// func moveZeroes(nums []int) {
// 	i := 0
// 	j := 0

// 	for i < len(nums)-1 && j < len(nums) {
// 		if nums[j] != 0 {
// 			nums[i], nums[j] = nums[j], nums[i]
// 			i += 1
// 			j += 1
// 		} else {
// 			j += 1
// 		}
// 	}

// }

// just use one pointer and iterate over array, overwrite end
func moveZeroes(nums []int) {
	i := 0
	for _, num := range nums {
		if num != 0 {
			nums[i] = num
			i += 1
		}
	}

	for ; i < len(nums); i++ {
		nums[i] = 0
	}
}
